Utility functions to scale model

Lars-DHondt-KUL opened the pull request:
<!--- Provide a general summary of your changes in the Title above -->

## Description
<!--- Provide a short overview of the changes -->
New utility functions to scale model parameters that are not adapted by the OpenSim scale tool.

## Motivation and Context
<!--- Why is this change required? What problem does it solve? -->
<!--- If it fixes an open issue, please link to the issue here. -->
OpenSim scale tool does not scale contact spheres nor maximal force of muscles, ligaments, actuators. _completeModelScaling.m_ performs all necessary scaling. It relies on subfunctions, which can be used independently to add, scale, or remove model components.

## How Has This Been Tested?
<!--- Please describe how you tested your changes. -->
<!--- Please describe the outcome of these tests. -->
Run _completeModelScaling.m_ and observe that this creates a fully scaled model.

## Suggested tests for reviewers
<!---Please describe tests for the reviewer to run -->
Try running _completeModelScaling.m_ for different models.

<!--- Assign labels ("Labels" tab in side bar). Each PR should have at least one "Review:..." label, since we use these to allocate reviewers. -->
<!--- Do not add requested reviewers, unless this person specifically agreed to review your PR. -->
<!--- If you are a collaborator (i.e. have direct write access to this repo) select yourself as assignee, otherwise someone will be assigned to your PR. -->

Lars-DHondt-KUL committed:
Create scaleMuscleForce.m

Lars-DHondt-KUL committed:
add utility functions

Lars-DHondt-KUL committed:
Create completeModelScaling.m

Lars-DHondt-KUL committed:
Update completeModelScaling.m

Lars-DHondt-KUL committed:
Update completeModelScaling.m

Lars-DHondt-KUL committed:
Update completeModelScaling.m

Lars-DHondt-KUL committed:
Merge branch 'master' into dev-modelscaling

Lars-DHondt-KUL committed:
Update completeModelScaling.m - citation

Lars-DHondt-KUL committed:
Merge branch 'master' into dev-modelscaling

InesVandekerckhove reviewed (changes requested):
- I get a model with double contact spheres: those from the original model and those that are scaled. Does removeContactSpheres.m not work?
<img width="305" height="720" alt="image" src="https://github.com/user-attachments/assets/099293a6-f353-4d66-9881-0030d7b0929d" />

The first contact spheres are from the original model; the ones with '_0' are the new, scaled contact spheres.
- The same issue with the actuators, I get double actuators; those from the original model and those that are scaled. Does removeActuators.m not work?
<img width="253" height="285" alt="image" src="https://github.com/user-attachments/assets/6c2d3c1f-3d88-4107-947d-eba11bbd98ba" />

- line 99: sf_contact_dissipation = 1/sf_length -> shouldn't it be -> = sf_length (without 1/) ?
- It is not clear to me what the difference is between the inputs 'osim_scaleTool' and 'osim_scale' ? Could you add more explanation here?
- Distinction between completeModelScaling.m and AdaptOpenSimModel.m not fully clear. Scaling the contact spheres and ligaments is now done in both scripts.
- typo line 59 -> 'calculate' instead of 'calculated'

Lars-DHondt-KUL committed:
contact parameter scaling

Scale contact parameters instead of removing contacts and adding new contacts with scaled parameters.

Lars-DHondt-KUL committed:
Merge branch 'master' into dev-modelscaling

Lars-DHondt-KUL committed:
remove scaling from AdaptOpenSimModel

Lars-DHondt-KUL committed:
get_contact_spheres -> getContactSpheres

Lars-DHondt-KUL committed:
correct mistakes in comments

Lars-DHondt-KUL committed:
Delete fixContactSpherePositionAfterScaling.m

function is hardcoded and outdated

Lars-DHondt-KUL commented:
I made some changes. Most notably:

- Changed `completeModelScaling`: Scale contact parameters instead of removing contacts and adding new contacts with scaled parameters.
- Removed scaling from `AdaptOpenSimModel`.

Lars-DHondt-KUL committed:
Update completeModelScaling.m

ellisvancan commented:
I tested the code with gait1018 and DHondt_4seg, generic and with different masses.

1. same issue with actuator as Ines describes above: 
<img width="300" height="337" alt="image" src="https://github.com/user-attachments/assets/0e5f5b0e-ee29-41f3-b748-da0b40e277b5" />
Contactspheres are no longer duplicated, so that fix worked

2.  When osim_scaled = osim_scaleTool,  I would like to have the option to automatically add a suffix such as _AdaptedforPredsim and _sf to the model name. This helps prevent accidentally scaling the same model multiple times without noticing. If you agree this would be useful, I can integrate this into the code.

ellisvancan commented on AdaptOpenSimModel/completeModelScaling.m:
```diff
@@ -0,0 +1,150 @@
+% Original date: 3 January 2025
+% --------------------------------------------------------------------------
+
+arguments
```
im not sure if this current setup fully serves its purpose. Since inputs are still position-dependent, it’s not possible to skip intermediate optional arguments (e.g., height_generic) while specifying later ones like the foot scale factors.

Alternative:
```
function [] = completeModelScaling(osim_scaleTool, osim_generic, scaling_settings)

arguments
    osim_scaleTool
    osim_generic
    scaling_settings.osim_scaled = osim_scaleTool;
    scaling_settings.height_subject = 1.65;
    scaling_settings.height_generic = 1.65;
    scaling_settings.scale_factor_foot_left = 1;
    scaling_settings.scale_factor_foot_right  = scaling_settings.scale_factor_foot_left;
end
```

ellisvancan commented on AdaptOpenSimModel/removeActuators.m:
```diff
@@ -0,0 +1,36 @@
+    if strcmp(force_j.getConcreteClassName(),'ActivationCoordinateActuator')
+        force_j.delete();
+    end
+end
```
<img width="520" height="466" alt="Image" src="https://github.com/user-attachments/assets/f83d2fd2-e580-4320-a257-10d99577b35f" />

when i try to interact with the model matlab keeps on crashing

ellisvancan replied:
What I tried:
```
%% check before
n_act_before = 0;
for j = 1:model.getForceSet().getSize()
    force_j = model.getForceSet().get(j-1);
    if strcmp(char(force_j.getConcreteClassName()), 'ActivationCoordinateActuator')
        n_act_before = n_act_before + 1;
    end
end

disp(['Actuators before: ', num2str(n_act_before)])

%% loop over forces
for j=1:model.getForceSet().getSize()
    force_j = model.getForceSet().get(j-1);
    if strcmp(force_j.getConcreteClassName(),'ActivationCoordinateActuator')
        force_j.delete();
        disp(force_j.getConcreteClassName())
    end
end


%% check after
n_act_after = 0;
for j = 1:model.getForceSet().getSize()
    force_j = model.getForceSet().get(j-1);
    if strcmp(char(force_j.getConcreteClassName()), 'ActivationCoordinateActuator')
        n_act_after = n_act_after + 1;
    end
end

disp(['Actuators after: ', num2str(n_act_after)])
```

ellisvancan reviewed (changes requested):
see comments above

ellisvancan committed:
integrating changes PR #243

- Name value pair options
- Warning when model name indicates that model has already been scaled

ellisvancan committed:
Update ModelPreparation.md

adding documentation

ellisvancan committed:
Update ModelPreparation.md

ellisvancan committed:
scaling stiffness and dissipation optional
